### ISSUE TITLE
[bitnami/redis] errors unmatched pid/can't find pid in our scripts execution queue

### ISSUE THREAD

**Issue by HighWatersDev** [bitnami/charts#7431]
**Which chart**:
Redis 15.3.1

**Describe the bug**
I see a lot of errors in redis and sentinel containers about process IDs.

In redis container:

```
1:M 07 Sep 2021 16:25:47.400 # Warning, detected child with unmatched pid: 2567
1:M 07 Sep 2021 16:25:47.500 # Warning, detected child with unmatched pid: 2584
1:M 07 Sep 2021 16:25:47.601 # Warning, detected child with unmatched pid: 2600
1:M 07 Sep 2021 16:25:47.701 # Warning, detected child with unmatched pid: 2616
```

In sentinel container:

```
1:X 07 Sep 2021 19:07:12.282 # waitpid() returned a pid (3306) we can't find in our scripts execution queue!
1:X 07 Sep 2021 19:07:17.163 # waitpid() returned a pid (3324) we can't find in our scripts execution queue!
1:X 07 Sep 2021 19:07:21.981 # waitpid() returned a pid (3340) we can't find in our scripts execution queue!
```

Consequently, each pod's readiness probes fail.

**To Reproduce**

1. Deploy Helm chart with sentinel and tls enabled
2. Wait for pods to provision
3. Open kubectl logs for each container

**Expected behavior**
I expect the service to remain stable and readiness probes to not fail every time it runs.

**Version of Helm and Kubernetes**:

- Output of `helm version`:

```
3.6.1
```

- Output of `kubectl version`:

```
1.20.7
```

**Comment by alemorcuq** [comment]
I'm not able to reproduce your issue with chart 15.3.2. Can you share your custom values to deploy and try to reproduce it?

**Comment by HighWatersDev** [comment]
These are the custom values I provided. I initially had this issue when using TLS [`#7392`](https://github.com/bitnami/charts/issues/7392#issue-986928168) which I'm still experiencing. So, I manually updated master address in sentinel config to allow sentinels to communicate. And those PID error is what I see now.

```
architecture: replication

auth:
  enabled: true
  sentinel: true
  password: "password"

master:
  livenessProbe:
    enabled: true
    initialDelaySeconds: 30
  readinessProbe:
    enabled: true
    initialDelaySeconds: 15
  resources:
    limits:
      memory: "500Mi"
      cpu: "0.2"
    requests:
      memory: "200Mi"
      cpu: "0.1"
  podAntiAffinityPreset: soft

replica:
  replicaCount: 3
  resources:
    limits:
      memory: "350Mi"
      cpu: "0.02"
    requests:
      memory: "200Mi"
      cpu: "0.01"

sentinel:
  enabled: true
  masterSet: masterset
  livenessProbe:
    enabled: true
  readinessProbe:
    enabled: true
  resources:
    limits:
      memory: "350Mi"
      cpu: "0.02"
    requests:
      memory: "200Mi"
      cpu: "0.01"

pdb:
  create: true
  minAvailable: 1

tls:
  enabled: true
  authClients: true
  autoGenerated: true
```

**Comment by alemorcuq** [comment]
Hi, @HighWatersDev.

I've tried your values.yaml, but I'm getting a different error than yours. I get the first node to run but I get an error on the sentinel of the second node:

```
$ k get po
NAME           READY   STATUS             RESTARTS   AGE
redis-node-0   2/2     Running            0          12m
redis-node-1   1/2     CrashLoopBackOff   7          11m
```

```
$ k logs -f redis-node-1 -c sentinel
Warning: Using a password with '-a' or '-u' option on the command line interface may not be safe.

*** FATAL CONFIG FILE ERROR (Redis 6.2.5) ***
Reading the configuration file, at line 3
>>> 'sentinel monitor masterset   2'
Unrecognized sentinel configuration statement.
```

I will try in a different environment to see if I can reproduce your issue.

One question: I can see that your logs say "warning". Apart from those warnings, is the chart working properly?

**Comment by HighWatersDev** [comment]
@alemorcuq you are correct, when I ran the above, I did run into this issue #7392. What I did to try to work around that is to manually set config files for sentinels. That's how I got the above warnings.

**Comment by alemorcuq** [comment]
Thanks. I'm adding this information to the task created in #7392 so we can investigate it after fixing the original issue.

**Comment by HighWatersDev** [comment]
Hi @alemorcuq,

Wonder if there is any update on this. I recently deployed redis-ha without tls enabled and am still experiencing the same issue `# waitpid() returned a pid (3340) we can't find in our scripts execution queue!`. It is only happening on sentinel containers.

**Comment by alemorcuq** [comment]
No, @HighWatersDev, we haven't had the chance to start working on this yet.

**Comment by jack1902** [comment]
Getting `Unrecognized sentinel configuration statement.` with values:

```yaml
auth:
  enabled: false

volumePermissions:
  enabled: true

sysctl:
  enabled: true

master:
  persistence:
    size: 10Gi
  livenessProbe:
    initialDelaySeconds: 120
  readinessProbe:
    initialDelaySeconds: 5

replica:
  persistence:
    size: 10Gi
  livenessProbe:
    initialDelaySeconds: 120
  readinessProbe:
    initialDelaySeconds: 5

sentinel:
  enabled: true
  failoverTimeout: 18000
  cleanDelaySeconds: 5
  livenessProbe:
    initialDelaySeconds: 120
  readinessProbe:
    initialDelaySeconds: 30

pdb:
  create: true
  minAvailable: 2
```

To add, I'm running on microk8s with `dns` and `storage` enabled

**Comment by migruiz4** [comment]
Hi @jack1902,

This discussion continues on this issue: https://github.com/bitnami/charts/issues/7392

**Comment by jwsomis** [comment]
I am seeing this same issue without TLS.

**Comment by ivan-kzn** [comment]
@jwsomis have you found the reason of the issue above?

**Comment by Dunge** [comment]
Same here, microk8s 1.26.4, just a sporadic error. HA cluster with bitnami chart

Kubernetes events on redis-node-1:
```
Readiness probe failed: command "sh -c /health/ping_sentinel.sh 5" timed out
```

Log on redis-node-1 sentinel container
```
1:X 13 May 2023 11:00:39.656 # +tilt #tilt mode entered
1:X 13 May 2023 11:00:40.682 # waitpid() returned a pid (1421280) we can't find in our scripts execution queue!
1:X 13 May 2023 11:01:09.675 # -tilt #tilt mode exited
```

**Comment by ieugen** [comment]
We are also seeing this error quite a lot in our cluster.

```
1:X 15 May 2023 08:03:58.443 # -tilt #tilt mode exited
1:X 15 May 2023 08:06:23.531 # +tilt #tilt mode entered
1:X 15 May 2023 08:06:23.531 # waitpid() returned a pid (930692) we can't find in our scripts execution queue!
1:X 15 May 2023 08:06:23.531 # waitpid() returned a pid (930701) we can't find in our scripts execution queue!
1:X 15 May 2023 08:06:33.898 # +tilt #tilt mode entered
1:X 15 May 2023 08:06:33.898 # waitpid() returned a pid (930683) we can't find in our scripts execution queue!
1:X 15 May 2023 08:06:33.898 # waitpid() returned a pid (930710) we can't find in our scripts execution queue!
1:X 15 May 2023 08:06:33.898 # waitpid() returned a pid (930728) we can't find in our scripts execution queue!
1:X 15 May 2023 08:06:33.898 # waitpid() returned a pid (930737) we can't find in our scripts execution queue!
1:X 15 May 2023 08:06:33.898 # waitpid() returned a pid (930719) we can't find in our scripts execution queue!
1:X 15 May 2023 08:06:33.898 # waitpid() returned a pid (930746) we can't find in our scripts execution queue!
1:X 15 May 2023 08:06:33.898 # waitpid() returned a pid (930764) we can't find in our scripts execution queue!
1:X 15 May 2023 08:06:33.898 # waitpid() returned a pid (930773) we can't find in our scripts execution queue!
1:X 15 May 2023 08:06:33.898 # waitpid() returned a pid (930755) we can't find in our scripts execution queue!
1:X 15 May 2023 08:07:03.954 # -tilt #tilt mode exited
1:X 15 May 2023 08:20:08.181 # +tilt #tilt mode entered
```

These are the customisations we are applying:

```
architecture: replication

# https://github.com/bitnami/charts/tree/main/bitnami/redis#redis-replicas-configuration-parameters
replica:
  replicaCount: 3
  persistence:
    enabled: true
    storageClass: retain-local-path

# https://github.com/bitnami/charts/tree/main/bitnami/redis#redis-sentinel-configuration-parameters
sentinel:
  enabled: true
  masterSet: XXXX
  persistence:
    enabled: true
    storageClass: retain-local-path

# https://github.com/bitnami/charts/tree/main/bitnami/redis#metrics-parameters
metrics:
  enabled: true
```

**Comment by Dunge** [comment]
Upgrading my chart from 17.0.11 to 17.10.3 and setting `useHostnames: false` as mentioned in #9689 seems to have helped. At least I don't have tilt mode coming anymore, so no waitpid() error in between, and no more readiness probe failing. For now, let's hope it remains like this.

Would still be preferable if we didn't have to set a custom config the change the default behaviour not to get errors. I don't quite grasp why does using hostnames vs ip would introduce time drift?

**Comment by smolinari** [comment]
We also ran into this problem today upgrading to 17.11.3. And as @Dunge setting `useHostnames: false` seems to have resolved the problem. Would be nice to have this remedied.

Scott

**Comment by migruiz4** [comment]
Hi there,

I have been investigating the issues with Redis Sentinel entering `tilt mode`, although we are not able to reproduce it locally in a dev environment.

As mentioned in this thread, the issue appears to be related to using hostnames to configure sentinel but not directly caused by it.
We suspect the issue could be caused by some I/O delay depending on the storage system used as mentioned [here](https://groups.google.com/g/redis-db/c/cul6TFUuKEE), and the issue becomes more severe if hostname usage is enabled.

From [Sentinel source code](https://github.com/redis/redis/blob/7.2/src/sentinel.c#L5357-L5386):
```
/* This function checks if we need to enter the TILT mode.
 *
 * The TILT mode is entered if we detect that between two invocations of the
 * timer interrupt, a negative amount of time, or too much time has passed.
 * Note that we expect that more or less just 100 milliseconds will pass
 * if everything is fine. However we'll see a negative number or a
 * difference bigger than SENTINEL_TILT_TRIGGER milliseconds if one of the
 * following conditions happen:
 *
 * 1) The Sentinel process for some time is blocked, for every kind of
 * random reason: the load is huge, the computer was frozen for some time
 * in I/O or alike, the process was stopped by a signal. Everything.
 * 2) The system clock was altered significantly.
 *
 * Under both this conditions we'll see everything as timed out and failing
 * without good reasons. Instead we enter the TILT mode and wait
 * for SENTINEL_TILT_PERIOD to elapse before starting to act again.
 *
 * During TILT time we still collect information, we just do not act. */
void sentinelCheckTiltCondition(void) {
    mstime_t now = mstime();
    mstime_t delta = now - sentinel.previous_time;


    if (delta < 0 || delta > sentinel_tilt_trigger) {
        sentinel.tilt = 1;
        sentinel.tilt_start_time = mstime();
        sentinelEvent(LL_WARNING,"+tilt",NULL,"#tilt mode entered");
    }
    sentinel.previous_time = mstime();
}
```

As mentioned in the comment, Tilt mode can be entered in several cases: Huge load, I/O delays, the process was stopped by a signal, etc.
Anything that delays Sentinel computation capacity will trigger it into tilt mode.

As it is also mentioned in the comment:
>  /* During TILT time we still collect information, we just do not act. */

That means that, while Sentinel continues working on Tilt mode, it won't be responsive to clients.

To solve this issue, there would be several solutions:
- Modify the livenessProbe timer so Sentinel container is not restarted during the `SENTINEL_TILT_PERIOD`. Currently liveness will restart the pod if unavailable for 25 seconds, while the default `tilt-period` is 30 seconds.
- Increase the tilt timer count by setting `tilt-trigger` in the `sentinel.configuration` value.
- Increase Sentinel container resources using `sentinel.resources` and/or check for delays in I/O operations in your storage system.

We will start by implementing the first option, incrementing the livenessProbe timeout from 25 to 60 seconds to not interrupt tilt mode.
Nonetheless, this won't prevent Sentinel from running into tilt-mode. 
To prevent tilt-mode, you may have to increase the resources of the sentinel container or check for issues that could be causing I/O delays with your storage class. Alternatively, you could increase the `tilt-trigger` to be more permissive with sentinel temporarily freezing, although we do not recommend it for production environments.

**Comment by pascal-hofmann** [comment]
See https://github.com/bitnami/charts/issues/10002#issuecomment-2482110450 for details about the message `waitpid() returned a pid (...) we can't find in our scripts execution queue!`.

**Comment by Rishab-shivhare** [comment]
in values.yaml

 master:
    shareProcessNamespace: true
